Cleanup of throws and asserts of goto_convert.cpp

NlightNFotis opened the pull request:
In the same spirit as #2904 I am taking over development of #2814 because the original repository is now defunct, and I can't go further with that PR. This PR will host further development of that line of work.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] My contribution is formatted in line with CODING_STANDARD.md.
- [n/a] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [n/a] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Convert throws and asserts to INVARIANTs for goto_convert.cpp

tautschnig reviewed (changes requested):
At the current state: please see https://github.com/diffblue/cbmc/pull/2814 for a number of blocking request.

Commit:
Substitute duplicate invariants with transfer function calls

Remove duplicate invariants, and change some to transfer (cast)
function calls, which provide guarantees about the data types
by construction. Also substitute invariants with exceptions
where appropriate.

Commit:
Refactor goto_convertt::case_guard

Refactor a function in goto_convertt so that it becomes
more comprehensible.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: a6c16991dc331f1cba2c9633753a11f93ac46180).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85897734
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 3e57a9c8eb4149dd2190f87b115ec68f07eac0ea).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85899011

NlightNFotis committed:
Change data_invariants to invariants_with_diagnostics

Change some data_invariants to invariants_with_diagnostics
where they are more appropriate, and remove some duplicate
invariants, where they are guaranteed by construction of the
types themselves.

NlightNFotis commented:
@tautschnig This should be now in a state where it's ready for a review. Just a note:

I have tried my best to make the commits as close as possible to the contribution guidelines. It's a little bit hairy, because 3 people have worked on this thus far, and one of them is no longer with us, but we have to be careful not to lose commit information and attribution information, hence why the commits are the way they are and I can not squash them better.

tautschnig commented on regression/ansi-c/goto_convert/switch_range_bounds/test.desc:
```diff
@@ -0,0 +1,8 @@
+CORE
```
Isn't this one GCC-only?

NlightNFotis replied:
No, both `clang` and `gcc` report the same error.

tautschnig replied:
Ok, but Visual Studio might not? But either way this one should fail in the front-end, so it's probably ok.

tautschnig commented on regression/ansi-c/goto_convert/invalid_goto_label/test.desc:
```diff
@@ -0,0 +1,8 @@
+CORE gcc-only
```
Why is this gcc-only?

NlightNFotis replied:
You're right, this is a mistake, both `clang` and `gcc` report the same error.

tautschnig commented on regression/ansi-c/goto_convert/switch_range_empty/test.desc:
```diff
@@ -0,0 +1,8 @@
+CORE
+main.c
+
+^EXIT=134$
```
Huh? That's not what should be happening...

NlightNFotis replied:
What should be happening? I see that `gcc` and `clang` produce warnings that the range is empty, but are happy to compile nonetheless. Should I remove the invariant?

tautschnig replied:
We should be able to support whatever GCC and Clang do. 134 (a failing invariant) is not an exit code we should ever be producing for a syntactically valid C program.

tautschnig commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -602,14 +573,7 @@ void goto_convertt::convert_expression(
-  }
-
-  exprt expr=code.op0();
+  exprt expr = code.expression();
```
Just curious: why can't this be a const reference?

NlightNFotis replied:
To my understanding it gets modified in the alternative branch of the if statement, by `clean_expr` and the rest of the code.

tautschnig commented on src/goto-programs/goto_convert.cpp:
```diff
@@ -1924,14 +1856,12 @@ irep_idt goto_convertt::get_string_constant(const exprt &expr)
-
-    throw 0;
-  }
+  bool res = get_string_constant(expr, result);
```
`const`

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 9a6e16fadcec4108da63658f54fc58c3b0f41dc8).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86143603

NlightNFotis committed:
Provide tests for goto_convert

NlightNFotis commented:
@tautschnig I decided on an alternative approach to the bugged test (the one we had a divergent behaviour from `gcc` and `clang`).

Before the refactoring work, that turned a line in `goto_convertt::convert_switch` in line 1252 from an `assert` to a `DATA_INVARIANT`. It seems that both of these behaviours are wrong to begin with, and I consider this a bug. For this reason I added a `KNOWNBUG` test, that I will mark as `CORE` in another PR, that I will submit on top of this PR fixing this, and in which I will only have the removal of the assertion and the activation of the test. So consider this one at a reviewable state.

NlightNFotis commented:
I pushed #3055 that fixes this the issue suggested above. That PR also for now contains the changes introduced here, but once this one is merged, I will rebase that one appropriately to only mirror the change it introduces.

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: 5e7e55ad8902b8873f910d5b88e36bc56bc1aefc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86177330
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: 552381869c6f82102a7233ce69b951e92d8b8cb6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86177840
